cloverage/src/cloverage/instrument.clj

```diff
@@ -79,59 +93,79 @@
+                 ;; namespace-less symbols
+                 {'#{try}         :try  ; try has to special case catch/finally
+                  '#{if do throw} :do   ; these special forms can recurse on all args
+                  '#{let* loop*}  :let
+                  '#{def}         :def  ; def can recurse on initialization expr
+                  '#{fn*}         :fn
+                  '#{set!}        :set  ; set must not evaluate the target expr
+                  '#{.}           :dotjava
+                  '#{case*}       :case*
+                  '#{new}         :new
+                  '#{reify*}      :reify*
+                  ;; FIXME: monitor-enter monitor-exit
+
+                  ;; namespaced macros
+                  `#{cond}        :cond ; special case cond to avoid false partial
+                  `#{loop let}    :let
+                  `#{letfn}       :letfn
+                  `#{for doseq}   :for
+                  `#{fn}          :fn
+                  `#{defn}        :defn     ; don't expand defn to preserve stack traces
+                  `#{defmulti}    :defmulti ; special case defmulti to avoid boilerplate
+                  `#{defprotocol} :atomic   ; no code in protocols
+                  `#{defrecord}   :record
+                  `#{deftype*}    :deftype*
+                  `#{ns}          :atomic
+
+                  ;; http://dev.clojure.org/jira/browse/CLJ-1330 means AOT-compiled definlines
+                  ;; are broken when used indirectly. Work around - do not wrap the definline
+                  `#{booleans bytes chars shorts floats ints doubles longs} :definlined-fn-call}
+                 symb symbols]
+             [symb typ])))
+
+(defn- list-type
+  "Return the type of a list or list-like (presumably a function/macro call) form.
+
+    (list-type '(= 1 2)) ; -> :list
+    (list-type '(fn [])) ; -> :fn"
+  ([[head :as form]]
+   (let [head (maybe-resolve-symbol head)]
+     (or (list-head-symbol->type head)
+         (cond
+           ;; A call to a function such as `clojure.core/int` that is defined normally with `defn` but has a special
+           ;; `:inline` version. Not necessarily inlined for all arities -- for example `+` is only inlined for 2+ args
+           (inlined-fn-call? form)  :inlined-fn-call
+           ;; A special form such as `quote` or `import*`
+           (atomic-special? head)   :atomic
+           ;; XXX: we used to not do anything with unknown specials, now we wrap them
+           ;; in a macro, then macroexpand back to original form. Methinks it's ok.
+           (special-symbol? head)   :unknown
+           :else                    :list))))
+
+  ([[head :as form] env]
+   ;; if the list is a call to a symbol with a local binding, but not a special form such as `var` or `new` (which
+   ;; aren't overshadowed by local variables) return `:list` type.
+   (if (and (get env head)
+            (not (special-symbol? head)))
+     :list
+     (list-type form))))
```

Comment:
Because `inlined-fn-call?` needs to look at the whole form I had to rework `list-type` to take the whole form instead of just the form at its head. As part of that I converted replaced the long `condp` call with a map of symbol -> type that speeds instrumentation up quite a bit.